modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DefaultImplicitConnectionProviderFactory.java

```diff
@@ -12,47 +12,67 @@
 import static org.mule.runtime.module.extension.internal.loader.utils.ImplicitObjectUtils.getFirstImplicit;
 import static org.mule.runtime.module.extension.internal.util.MuleExtensionUtils.getAllConnectionProviders;
 import org.mule.runtime.api.connection.ConnectionProvider;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.exception.MuleRuntimeException;
 import org.mule.runtime.api.meta.model.ExtensionModel;
 import org.mule.runtime.api.meta.model.config.ConfigurationModel;
 import org.mule.runtime.api.meta.model.connection.ConnectionProviderModel;
 import org.mule.runtime.core.api.Event;
 import org.mule.runtime.core.api.MuleContext;
-import org.mule.runtime.api.exception.MuleException;
-import org.mule.runtime.api.exception.MuleRuntimeException;
 import org.mule.runtime.core.internal.connection.ConnectionManagerAdapter;
 import org.mule.runtime.module.extension.internal.runtime.resolver.ResolverSet;
 
+import java.util.Optional;
+
+import javax.inject.Provider;
+
 /**
  * Default implementation of {@link ImplicitConnectionProviderFactory}
  *
  * @since 4.0
  */
-public final class DefaultImplicitConnectionProviderFactory implements ImplicitConnectionProviderFactory {
+public final class DefaultImplicitConnectionProviderFactory<T> implements ImplicitConnectionProviderFactory {
 
-  ConnectionManagerAdapter connectionManagerAdapter;
+  private final ExtensionModel extensionModel;
+  private final MuleContext muleContext;
+  private final Provider<ResolverSet> resolverSetProvider;
+  private ConnectionManagerAdapter connectionManagerAdapter = null;
+  private ConnectionProviderModel connectionProviderModel = null;
+  private ResolverSet resolverSet = null;
+
+  public DefaultImplicitConnectionProviderFactory(ExtensionModel extensionModel, ConfigurationModel configurationModel,
+                                                  MuleContext muleContext) {
+    this.extensionModel = extensionModel;
+    this.muleContext = muleContext;
+
+    resolverSetProvider = () -> {
+      synchronized (this) {
+        if (resolverSet == null) {
+          connectionProviderModel = getFirstImplicit(getAllConnectionProviders(extensionModel, configurationModel));
+
+          if (connectionProviderModel == null) {
+            throw new IllegalStateException(format(
+                                                   "Configuration '%s' of extension '%s' does not define a connection provider and none can be created automatically. Please define one.",
+                                                   configurationModel.getName(), configurationModel.getName()));
+          }
+          resolverSet = buildImplicitResolverSet(connectionProviderModel, muleContext);
```

Comment:
after this line, add `resolverSetProvider = () -> resolverSet so that the synchronized block doesn't add no longer needed contention